db/version_edit.cc

```diff
@@ -55,6 +55,8 @@ enum Tag : uint32_t {
   kDbId,
   kBlobFileAddition,
   kBlobFileGarbage,
+  kWalAddition,
```

Comment:
The WAL related edits are designed to track on on-disk WAL states. So WalAddition means there is a WAL on-disk, the WalAddition may have optional size / checksum, it's not related to logical WAL operations.

On consistency checks, the WAL related edits are designed to verify on-disk WALs on recovery, the logical checks such as must be created before being closed are not the design purpose.

Reply:
Tracking the logically obsolete WAL files is the same as tracking the `MinLogNumber` I mentioned in the PR summary. This `MinLogNumber` is already tracked in MANIFEST during every flush. But for unflushed empty column families, their min log number is always the first log number, so `MinLogNumber` is only a lower bound of the logs that should be kept alive.

I prefer viewing the WAL as the source of truth for logical data including the in-memory unflushed data, and MANIFEST as  the source of truth for on-disk files. For example, MANIFEST tracks SST files. In this PR, we make it track WAL files. In future PRs, maybe we let it track option files.

Reply:
> Tracking the logically obsolete WAL files is the same as tracking the `MinLogNumber` I mentioned in the PR summary. This `MinLogNumber` is already tracked in MANIFEST during every flush. But for unflushed empty column families, their min log number is always the first log number, so `MinLogNumber` is only a lower bound of the logs that should be kept alive.
> 

Hm, I don't think they're the same thing. I do understand the corner case with empty column families. My point is that we already have a way (used in `FindObsoleteFiles`) of identifying the actual set of obsolete WAL files, as opposed to just a soft lower bound, right? IIUC, your plan is to log a dedicated `VersionEdit` (with `DeleteWal`) at that point. What I'm suggesting is moving this up in time a bit and piggybacking that same `DeleteWal` on the `VersionEdit` that is logged for the (last) flush that makes the WAL file obsolete. Please let me know if I'm missing something.

Reply:
Are you suggesting piggybacking the WalDeletion in https://github.com/facebook/rocksdb/blob/master/db/flush_job.cc#L187 ?
Yes, since db mutex is held when entering this method, we can delete WALs in WalSet whose log number < GetNextLogNumber. One complexity here is that in 2pc mode, this is not the logic for determining min log number, see https://github.com/facebook/rocksdb/blob/master/db/db_impl/db_impl_files.cc#L25

Reply:
Yes, making the `DeleteWal`s a part of the `FlushJob`'s `VersionEdit` is what I was thinking. Of course, we would need to make sure that both the non-2PC and the 2PC case are handled consistently with what `FindObsoleteFiles` does.

Reply:
Since this PR does not implement wal deletion anyway, we can continue this discussion in a future PR dedicated to wal deletion. I'll try to make logical wal deletions in flush jobs. In this PR, I removed assumption WalDeletion cannot be part of other version edits. See the last commit.